File: app/customer/extraction_rules.py
"""Dependency-cascade rules for HubSpot object toggles.

When one toggle is disabled, certain other toggles become impossible
(`lead_pipelines` is meaningless without `leads`, `form_submissions` without
`forms`, etc.). The frontend uses these to preview and grey-out affected rows;
the backend re-applies them on save so a bad PUT can't bypass the rules.

`OBJECT_GROUPS` is consumed by the frontend ObjectToggleGrid to render the
hybrid grouped-by-default / per-type-on-expand UX.
"""

from __future__ import annotations

import copy


# Hybrid grouping used by the frontend toggle grid.
OBJECT_GROUPS: dict[str, dict] = {
    "CRM": {
        "children": ["contacts", "companies", "deals", "leads"],
        "expandable": False,
    },
    "Activities": {
        "children": ["calls", "meetings", "emails", "notes", "tasks"],
        "expandable": True,  # rendered as a single "Activities" group, expandable
        "container_key": "activities",
    },
    "Marketing": {
        "children": ["campaigns", "forms", "form_submissions"],
        "expandable": False,
    },
    "Other": {
        "children": ["owners", "deal_pipelines", "lead_pipelines"],
        "expandable": False,
    },
}


# Hard dependencies: if `key` is disabled, every `forced_off` toggle must also
# be disabled. The frontend uses this for live preview; the backend applies it
# on save.
DEPENDENCIES: dict[str, list[str]] = {
    "leads":    ["lead_pipelines"],
    "forms":    ["form_submissions"],
    "deals":    [],
    "contacts": [],
    "companies": [],
}


def apply_cascade(toggles: dict) -> dict:
    """Return a new toggle dict with dependent toggles forced-off where needed.

    The input is a mutable nested dict matching `DEFAULT_OBJECTS` in extraction.py.
    """
    out = copy.deepcopy(toggles)

    # First pass: handle the activities sub-dict (no cascade needed inside,
    # but normalize structure if missing).
    if "activities" not in out or not isinstance(out["activities"], dict):
        out["activities"] = {
            "calls": True, "meetings": True, "emails": True, "notes": True, "tasks": True,
        }

    # Second pass: hard cascades.
    changed = True
    while changed:
        changed = False
        for trigger, forced in DEPENDENCIES.items():
            if out.get(trigger, True) is False:
                for off in forced:
                    if out.get(off, True) is not False:
                        out[off] = False
                        changed = True

    return out


def describe_cascade(toggles: dict) -> dict[str, list[str]]:
    """Return a map: toggle key → list of other toggles it would force off if
    the user disabled it. Used by the frontend to render the tooltip preview.
    """
    out: dict[str, list[str]] = {}
    for trigger, forced in DEPENDENCIES.items():
        if forced:
            out[trigger] = list(forced)
    return out

File: frontend/src/lib/extractionRules.ts
/**
 * TS mirror of app/customer/extraction_rules.py. Keep in sync.
 * Used by the ObjectToggleGrid for live cascade preview.
 */

export interface ObjectsState {
  contacts: boolean;
  companies: boolean;
  deals: boolean;
  leads: boolean;
  owners: boolean;
  deal_pipelines: boolean;
  lead_pipelines: boolean;
  activities: {
    calls: boolean;
    meetings: boolean;
    emails: boolean;
    notes: boolean;
    tasks: boolean;
  };
  campaigns: boolean;
  forms: boolean;
  form_submissions: boolean;
}

export const DEFAULT_OBJECTS: ObjectsState = {
  contacts: true,
  companies: true,
  deals: true,
  leads: true,
  owners: true,
  deal_pipelines: true,
  lead_pipelines: true,
  activities: { calls: true, meetings: true, emails: true, notes: true, tasks: true },
  campaigns: true,
  forms: true,
  form_submissions: true,
};

export const DEPENDENCIES: Record<string, string[]> = {
  leads: ["lead_pipelines"],
  forms: ["form_submissions"],
};

export function applyCascade(input: ObjectsState): ObjectsState {
  const out: ObjectsState = JSON.parse(JSON.stringify(input));
  if (!out.activities) {
    out.activities = { calls: true, meetings: true, emails: true, notes: true, tasks: true };
  }
  let changed = true;
  while (changed) {
    changed = false;
    for (const trigger of Object.keys(DEPENDENCIES)) {
      const t = trigger as keyof ObjectsState;
      if ((out[t] as boolean) === false) {
        for (const off of DEPENDENCIES[trigger]) {
          const o = off as keyof ObjectsState;
          if ((out[o] as boolean) !== false) {
            (out as any)[o] = false;
            changed = true;
          }
        }
      }
    }
  }
  return out;
}

export interface ObjectGroup {
  name: string;
  children: string[];
  expandable: boolean;
  containerKey?: keyof ObjectsState;
}

export const OBJECT_GROUPS: ObjectGroup[] = [
  { name: "CRM", children: ["contacts", "companies", "deals", "leads"], expandable: false },
  {
    name: "Activities",
    children: ["calls", "meetings", "emails", "notes", "tasks"],
    expandable: true,
    containerKey: "activities",
  },
  { name: "Marketing", children: ["campaigns", "forms", "form_submissions"], expandable: false },
  { name: "Other", children: ["owners", "deal_pipelines", "lead_pipelines"], expandable: false },
];

export function describeImpact(disabledKey: string): string[] {
  const out: string[] = [];
  const forced = DEPENDENCIES[disabledKey] || [];
  for (const f of forced) out.push(f);
  // Surface common downstream impacts for the tooltip
  const downstream: Record<string, string[]> = {
    leads: ["dim_leads", "bridge_lead_*", "agg_lead_health"],
    forms: ["fact_form_submissions"],
    deals: ["dim_deals", "agg_deal_*"],
    contacts: ["dim_contacts", "all bridge_*_contact"],
    companies: ["dim_companies", "all bridge_*_company"],
  };
  if (downstream[disabledKey]) {
    out.push(...downstream[disabledKey]);
  }
  return out;
}
